fhir-codegen/gradle-plugin/src/main/kotlin/com/google/fhir/codegen/EnumTypeSpecGenerator.kt

```diff
@@ -0,0 +1,288 @@
+                  .build()
+              )
+              .addProperty(
+                PropertySpec.builder(
+                    "definition",
+                    String::class.asClassName().copy(nullable = true),
+                    KModifier.PRIVATE,
+                  )
+                  .initializer("definition")
+                  .build()
+              )
+          }
+          addFunction(
+            FunSpec.builder("toString")
+              .addModifiers(KModifier.OVERRIDE)
+              .addStatement("return code")
+              .returns(String::class)
+              .build()
+          )
+          addFunction(
+            FunSpec.builder("getCode")
+              .addModifiers(KModifier.PUBLIC)
+              .returns(String::class)
+              .addStatement("return code")
+              .build()
+          )
+          addFunction(
+            FunSpec.builder("getSystem")
+              .addModifiers(KModifier.PUBLIC)
+              .returns(String::class)
+              .addStatement("return system")
+              .build()
+          )
+          addFunction(
+            FunSpec.builder("getDisplay")
+              .addModifiers(KModifier.PUBLIC)
+              .returns(String::class.asClassName().copy(nullable = true))
+              .addStatement("return display")
+              .build()
+          )
+          addFunction(
+            FunSpec.builder("getDefinition")
+              .addModifiers(KModifier.PUBLIC)
+              .returns(String::class.asClassName().copy(nullable = true))
+              .addStatement("return definition")
+              .build()
+          )
+          addType(
+            TypeSpec.companionObjectBuilder()
+              .addFunction(
+                FunSpec.builder("fromCode")
+                  .addParameter("code", String::class)
+                  .returns(ClassName.bestGuess(enumClassName))
+                  .beginControlFlow("return when (code)")
+                  .apply {
+                    fhirEnum.constants.forEach { addStatement("%S -> %L", it.code, it.name) }
+                    addStatement(
+                      "else -> throw IllegalArgumentException(\"Unknown code \$code for enum %L\")",
+                      enumClassName,
+                    )
+                  }
+                  .endControlFlow()
+                  .build()
+              )
+              .build()
+          )
+        }
+        .build()
+    return typeSpec
+  }
+
+  /**
+   * Instantiate a [FhirEnum] to facilitate the generation of Kotlin enum classes. The enum
+   * constants are derived from concepts defined in a `ValueSet`, a `CodeSystem`, or both. When both
+   * are provided, the constants are generated by intersecting the concepts from the `ValueSet` and
+   * the `CodeSystem`. If only one is provided, the constants are based solely on that source. Any
+   * nested concepts are flattened, transformed, and included in the resulting enum constants.
```

Comment:
can you explain why they are flatterend?

Reply:
The reason for flattening the nested concepts is that the nested codes will be used in the same context as the parent. This approach was borrowed from how HAPI structures handle nested concepts.